delfin/drivers/hpe/hpe_3par/component_handler.py

```diff
@@ -0,0 +1,243 @@
+                    elif (volume.get('state') == consts.STATUS_VOLUME_DEGRADED
+                          or volume.get('state') ==
+                          consts.STATUS_VOLUME_FAILED):
+                        status = constants.StoragePoolStatus.ABNORMAL
+
+                    # Get CPG name CPG name from which
+                    # the user space is allocated.
+                    # /CPG name from which the snapshot
+                    # (snap and admin) space is allocated.
+                    # userCPG snapCPG maybe not exist
+                    orig_pool_id = ''
+                    if 'userCPG' in volume:
+                        orig_pool_id = volume.get('userCPG')
+                    if 'snapCPG' in volume:
+                        if orig_pool_id == '':
+                            orig_pool_id = volume.get('snapCPG')
+                        else:
+                            orig_pool_id = orig_pool_id + '/' + \
+                                volume.get('snapCPG')
+
+                    compressed = True
+                    deduplicated = True
+
+                    vol_type = constants.VolumeType.THICK
+                    if volume.get('provisioningType') == consts.THIN_LUNTYPE:
+                        vol_type = constants.VolumeType.THIN
+
+                    # Virtual size of volume in MiB (10242bytes).
+                    usr_used = int(volume['userSpace']['reservedMiB']) * \
+                        consts.MiB_TO_Bytes
+                    admin_used = int(volume['adminSpace']['reservedMiB']) * \
+                        consts.MiB_TO_Bytes
+                    snap_used = int(volume['snapshotSpace']['reservedMiB']) \
+                        * consts.MiB_TO_Bytes
+                    total_cap = int(volume['sizeMiB']) * \
+                        consts.MiB_TO_Bytes
+                    used_cap = usr_used + admin_used + snap_used
+                    free_cap = total_cap - used_cap
+
+                    v = {
+                        'name': volume.get('name'),
+                        'storage_id': self.storage_id,
+                        'description': volume.get('comment'),  # none in 1.2
+                        'status': status,
+                        'native_volume_id': volume.get('id'),
+                        'native_storage_pool_id': orig_pool_id,
+                        'wwn': volume.get('wwn'),
+                        'type': vol_type,
+                        'total_capacity': total_cap,
+                        'used_capacity': used_cap,
+                        'free_capacity': free_cap,
+                        'compressed': compressed,
+                        'deduplicated': deduplicated
+                    }
+                    volume_list.append(v)
+            return volume_list
+
+        except Exception as err:
+            LOG.error(
+                "Failed to get list volumes from Hpe3parStor: {}".format(err))
+            raise exception.StorageBackendException(
```

Comment:
Is this exception correct for all value collection exceptions too? "get list volumes" maybe successful but parsing may have an issue